### DISCUSSION TITLE
Feature request - add support for hardware volume buttons = PgUp / PgDn (e.g. digital music stand)

### DISCUSSION THREAD

**Discussion by fizzybrain** [sumatrapdfreader/sumatrapdf#2699] [Ideas]
Thank you for SumatraPDF! It is exemplary in its efficiency and smallness.

I would like to use it with large-screen tablets to turn them into electronic music stands, using vol+, vol- hardware buttons.
Large-screen tablets tend to be really expensive.
However, large-screen tablets with non-functioning digitisers tend to be really cheap ;-)
Once a playlist has been created as a single PDF, all that is really required is Page Up and Page Down. Many tablets have hardware volume keys which would be ideal for this (the idea being that the setup happens before a gig with a bluetooth keyboard, say, and then the tablet is put on the stand and the PDF can be advanced essentially as a manual slideshow using those buttons).
I am please to see that keyboard shortcuts can now be customised, which is extremely promising. Looking at this list, though:
[https://github.com/sumatrapdfreader/sumatrapdf/blob/master/src/Accelerators.cpp#L14](https://github.com/sumatrapdfreader/sumatrapdf/blob/master/src/Accelerators.cpp#L14)
I can't identify anything that looks like volume buttons, unless I am missing something.
Maybe I should be able to work it out, but I imagine someone here could tell me if this is already possible (or not) far more quickly than I can do so.
(I could use autohotkey, with which I am vaguely familiar, but being able to bundle the PDF file with a portable SumatraPDF on a USB stick to play - and even auto-run - on any device without installing autohotkey would be nice ;-) )
Respect,
Ian

**Comment by GitHubRulesOK** [comment]
currently volume is outside SumatraPDF remit but those keys are usually tied to F2 & F3 keys so you could try ctrl or shift with those

AVIF is nearest to AV format but is stills only, thus situation could change if more audio formats included.

With AHK you should be able to map vol controls and if needed start ahk opening SumatraPDF plus your custom controls see my frozen https://github.com/GitHubRulesOK/SumatraPDF-Plus for an example

**Comment by kjk** [comment]
I've added `VolumeUp` and `VolumeDown` in https://github.com/sumatrapdfreader/sumatrapdf/commit/4e7e1a7520ed78b18be9ba0c78d8e5dc2762776d

Can't test those myself so let me know if those work.

If not, it's really a matter of figuring out which keys are those (http://www.kbdedit.com/manual/low_level_vk_list.html)

**Reply by fizzybrain** [reply]
That's amazing and great, thanks. :-)
I am away from suitable devices at the moment, but will test when I can.
(if I can work out how to compile things and do that. I'm more of a feature requester than feature tester, if you know what I mean)

**Reply by fizzybrain** [reply]
Finally got back home and got a device with Vol+/Vol- keys in my hands. Then downloaded Visual Studio, began looking up how to compile things and my nerve failed me. I do intend to learn this stuff, but not right now.
I'm happy to wait for 3.4.7 ;-)

**Reply by GitHubRulesOK** [reply]
@fizzybrain you should not need to compile simply download portable pre-release 3.5.X 
not list in master features ? but should be/was ? in https://www.sumatrapdfreader.org/prerelease

the cmd pair is added but you may need to match which keys the volume control adjusts 
V(VK_VOLUME_DOWN, "VolumeDown") \
V(VK_VOLUME_UP, "VolumeUp") 

without a separate control it is often in some ways related to a function triplet like F1 F2 F3 so on my keyboard those 3 keys are all too easy to brush against accidentally so using F1 for help it mutes and there is an loud expletive from me, or when attempting to use Function 3 for search and forgetting to select as function every thing gets quieter !! 

I tried setting those key on my keyboard without direct success in reassigning functions but I can if I use shift F keys which does not help you

Shortcuts [
[
		Cmd = CmdScrollUpHalfPage
		Key = VolumeUp
]
	[
		Cmd = CmdScrollDownHalfPage
		Key = VolumeDown
	]
]

**Reply by fizzybrain** [reply]
Yes, that works! Excellent! It still changes the volume, of course, but that doesn't matter for my purposes.
(tested with media remote control and a bluetooth keyboard with media controls which work when it's set to "android" mode)
+1 for the why-I-will-never-fully-convert-to-Linux list.
Many many thanks - I feel ... recognised.
Share and Enjoy,
Ian

**Reply by GitHubRulesOK** [reply]
> still changes the volume 🤣 

theoretically good its not blocking the function is it just 2% up down ?

@kjk would it be possible to include mute on/of toggle (VK_VOLUME_MUTE) for potential use with user choice of key before and after

**Reply by GitHubRulesOK** [reply]
@fizzybrain 
Actually that reminds me I wrote a reset volume Vol20.cmd for set audio to 20% use it raw but you can edit as needed to perhaps reset after a large number of up & downs, without audio prompt you dont need the speak or line after, simply down to zero then up to desired medium level.

```
@echo off
Title Volume=20

REM CHR(173) is the MUTE ON/OFF key, CHR(174) is the VOLUME DOWN key & CHR(175) is VOLUME UP key.
REM The next value after steps +/- 2% units, so UP 50 times will force volume to the max first.
REM In some cases depending on other apps it may be safer to set to Zero=(174)50 1st then up&down.

REM For example set 1st to ZERO (-50) up to 50% (+25) for notification then say down to 20% (-15)

(echo CreateObject("WScript.Shell"^).SendKeys "{" ^& chr(174^) ^&" 50}")>"%Temp%\%~n0.vbs"
(echo CreateObject("WScript.Shell"^).SendKeys "{" ^& chr(175^) ^&" 25}")>>"%Temp%\%~n0.vbs"
(echo CreateObject("SAPI.SpVoice"^).Speak ("Lowering volume to 20%%"^))>>"%Temp%\%~n0.vbs"
(echo CreateObject("WScript.Shell"^).SendKeys "{" ^& chr(174^) ^&" 15}")>>"%Temp%\%~n0.vbs"
cscript //NoLogo "%Temp%\%~n0.vbs"
```

**Reply by fizzybrain** [reply]
> > still changes the volume 🤣
> 
> theoretically good its not blocking the function is it just 2% up down ?
> 
> @kjk would it be possible to include mute on/of toggle (VK_VOLUME_MUTE) for potential use with user choice of key before and after

As I say, it doesn't matter for my purpose (digital music stand for live performance - I won't be playing audio from the device, but I dare say it could matter to others if they were using the device for accompaniment, say).
To answer your question, though - yes, 2% steps as normal.
Muted audio gets un-muted when volume key is pressed.
Maybe autohotkey can block Vol+/- (or send additional Vol-/+ to counteract) if it ever becomes a problem....
